diff --git a/dlls/quartz/memallocator.c b/dlls/quartz/memallocator.c
index 7a171769574..1e2f96ceaef 100644
--- a/dlls/quartz/memallocator.c
+++ b/dlls/quartz/memallocator.c
@@ -285,11 +285,11 @@ static HRESULT WINAPI BaseMemAllocator_Decommit(IMemAllocator * iface)
                 if (This->lWaiting != 0)
                     ERR("Waiting: %ld\n", This->lWaiting);
 
-                This->bCommitted = FALSE;
-                CloseHandle(This->hSemWaiting);
-                This->hSemWaiting = NULL;
+                // This->bCommitted = FALSE;
+                // CloseHandle(This->hSemWaiting);
+                // This->hSemWaiting = NULL;
 
-                hr = This->fnFree(iface);
+                // hr = This->fnFree(iface);
             }
         }
     }
@@ -386,13 +386,13 @@ static HRESULT WINAPI BaseMemAllocator_ReleaseBuffer(IMemAllocator * iface, IMed
             if (This->lWaiting != 0)
                 ERR("Waiting: %ld\n", This->lWaiting);
 
-            This->bCommitted = FALSE;
-            This->bDecommitQueued = FALSE;
+            // This->bCommitted = FALSE;
+            // This->bDecommitQueued = FALSE;
 
-            CloseHandle(This->hSemWaiting);
-            This->hSemWaiting = NULL;
+            // CloseHandle(This->hSemWaiting);
+            // This->hSemWaiting = NULL;
             
-            This->fnFree(iface);
+            // This->fnFree(iface);
         }
     }
     LeaveCriticalSection(This->pCritSect);
